File: pipeline/kite9-pipeline-common/src/main/java/com/kite9/k9server/command/xml/replace/ReplaceAttr.java
package com.kite9.k9server.command.xml.replace;

import org.kite9.diagram.dom.elements.ADLDocument;
import org.springframework.util.StringUtils;
import org.w3c.dom.Element;

import com.kite9.k9server.adl.holder.pipeline.ADLDom;

public class ReplaceAttr extends AbstractReplaceCommand<String, String> {

	public String name;
	
	@Override
	protected String getFromContent(ADLDom in) {
		return from;
	}

	@Override
	protected String getToContent(ADLDom in) {
		return to;
	}

	@Override
	protected String getExistingContent(ADLDom in) {
		Element on = findFragmentElement(in.getDocument(), fragmentId);
		return on.getAttribute(name);
	}

	@Override
	protected void doReplace(ADLDom existing, String site, String value, String old) {
		ADLDocument doc = existing.getDocument();
		Element e = findFragmentElement(doc, fragmentId);

		if (value == null) {
			e.removeAttribute(name);
		} else {
			e.setAttribute(name, value);
		}
	
		
		LOG.info("Processed replace attribute of "+fragmentId+" "+name);
	}

	@Override
	protected void checkProperties() {
		ensureNotNull("name", name);
		super.checkProperties();
	}

	@Override
	protected Mismatch same(String existing, String with) {
		if (StringUtils.isEmpty(existing) && StringUtils.isEmpty(with)) {
			return null;
		}
		
		return super.same(existing, with);
	}
	
	
	
	
}

File: pipeline/kite9-pipeline-common/src/main/kotlin/com/kite9/k9server/command/xml/replace/ReplaceAttr.kt
package com.kite9.k9server.command.xml.replace

import com.kite9.k9server.adl.holder.pipeline.ADLDom
import com.kite9.k9server.command.Command
import com.kite9.k9server.command.CommandContext
import org.w3c.dom.Document

class ReplaceAttr : AbstractReplaceCommand<String, String?>() {

    @JvmField
	var name: String? = null

    override fun getFromContent(d: ADLDom, ctx: CommandContext): String {
        return from!!
    }

    override fun getToContent(d: ADLDom, ctx: CommandContext): String {
        return to!!
    }

    override fun getExistingContent(o: ADLDom, ctx: CommandContext): String {
        val on = findFragmentElement(o.document, fragmentId!!, ctx)
        return on.getAttribute(name)
    }

    override fun doReplace(existing: ADLDom, site: String, value: String?, old: String?, ctx: CommandContext) {
        val doc: Document = existing.document
        val e = findFragmentElement(doc, fragmentId, ctx)
        if (value == null) {
            e.removeAttribute(name)
        } else {
            e.setAttribute(name, value)
        }
        ctx.log("Processed replace attribute of $fragmentId $name")
    }

    override fun checkProperties() {
        ensureNotNull("name", name)
        super.checkProperties()
    }

    override fun same(existing: String, with: String?): Command.Mismatch? {
        return if (isEmpty(existing) && isEmpty(with)) {
            null
        } else super.same(existing, with)
    }
}